TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+        /// The Current Program symbol being built as a Scope
+        /// </summary>
+        private Program CurrentProgram
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current entered node.
+        /// </summary>
+        private Node CurrentNode
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current basic block
+        /// </summary>
+        protected BasicBlockForNode CurrentBasicBlock
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current section symbol.
+        /// </summary>
+        internal CfgSectionSymbol CurrentSection
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current paragraph symbol.
+        /// </summary>
+        internal CfgParagraphSymbol CurrentParagraph
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current Sentence in the current Paragraph.
+        /// </summary>
+        internal CfgSentence CurrentSentence
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The Section and Paragraph Domain of this program.
+        /// </summary>
+        internal Container<Symbol> SectionsParagraphs
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Ordered list of all Sections an Paragraphs encountered in order.
+        /// </summary>
+        internal List<Symbol> AllSectionsParagraphs
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// All pending Goto instructions that will be handled at the end of the Procedure Division.
+        /// </summary>
+        protected LinkedList<Tuple<Goto, BasicBlockForNode>> PendingGOTOs;
+        /// <summary>
+        /// All encountered PERFORM procedure instructions
+        /// </summary>
+        protected LinkedList<Tuple<PerformProcedure, BasicBlockForNodeGroup>> PendingPERFORMProcedures;
+        /// <summary>
+        /// Pending ALTER instructions that will be handled at the end of the Procedure Division.
+        /// </summary>
+        protected LinkedList<Tuple<Alter, BasicBlockForNode>> PendingALTERs;
+
+        /// <summary>
+        /// More Symbol Reference associated to altered GOTO statement.
+        /// </summary>
+        protected Dictionary<Goto, HashSet<SymbolReference>> PendingAlteredGOTOS;
+
+        /// <summary>
+        /// All pending Next Sentence instructions that will be handled at the end of the Procedure Division.
+        /// </summary>
+        internal LinkedList<Tuple<NextSentence, BasicBlockForNode, CfgSentence>> PendingNextSentences;
+
+        /// <summary>
+        /// All encountered sentences
+        /// </summary>
+        internal List<CfgSentence> AllSentences;
+
+        public IList<Diagnostic> Diagnostics { get; private set; }
```

Comment:
Diagnostics are never read.
I think these Diagnostics should be returned by CompilationUnit.AllDiagnostics.
So if you activate CFG, you can keep your current code to check Diagnostics.
I know CompilationUnit is in TypeCobol project but we can find a way to "plug" CFG diagnostics into it.

Reply:
Yes this still a point to clarify how CFG diagnostics are reported or accessed by callers. So We discussed to have a CfgContext, which can be given a DiagnosticsDispatcher interface instance.